### ISSUE TITLE
Training instability: Periodic loss fluctuations and sinusoidal grad_norm patterns in Llama 3.2 (1B/500M) pre-training

### ISSUE THREAD

**Issue by ji-chengzhi** [NVIDIA-NeMo/Automodel#1679]
**Describe the bug**

We encountered significant training instability when pre-training Llama 3.2 (1B and 500M) models using the nemo_automodel framework.

Key Symptoms:

Loss Fluctuations: The loss starts to oscillate around step 1200.

<img width="2776" height="1304" alt="Image" src="https://github.com/user-attachments/assets/53196dff-152d-4510-9dc1-aceafeccc144" />

Sinusoidal Grad Norm: The grad_norm exhibits a distinct, periodic sinusoidal-like pattern rather than stabilizing.

<img width="2800" height="1304" alt="Image" src="https://github.com/user-attachments/assets/eba2de8a-56f1-4051-b022-646ac4ddd008" />

Scope of the Issue:

- Framework Versions: Reproduced on tag v0.2.0, v0.3.0, and the main branch.

- Model Scales: Observed in both 1B and 500M models.

- Datasets: Tested with multiple sources including [`EleutherAI/pythia_pile_idxmaps`](https://huggingface.co/datasets/EleutherAI/pythia_pile_idxmaps) (preprocessed bin/idx), `fineweb-edu` sample-100BT, and `Nemotron-CC-v2` High-Quality (processed by `tools/preprocess_megatron_dataset.py` with llama tokenizer). The issue persists regardless of the data source.

- Hardware: Reproduced across different GPU architectures (e.g., A800 * 8 nodes).

- Liger Kernel: We have explicitly tested with `use_liger_kernel: false`, and the issue still persists.

**Steps/Code to reproduce bug**

1. Use nemo_automodel 0.2.0 (also could reproduce on 0.3.0 and 0.3.0rc0).  

2. Use the provided YAML configuration below.

3. Initialize pre-training with provided Llama 3.2 1B `config.json` & tokenizer ([llama_1b_pythia.tar.gz](https://github.com/user-attachments/files/26474502/llama_1b_pythia.tar.gz)).

4. Monitor and Observe the breakdown of stability starting around step 1200.

```YAML
step_scheduler:
  global_batch_size: 1024
  local_batch_size: 8
  ckpt_every_steps: 2000
  val_every_steps: 2000 
  num_epochs: 1
  max_steps: 40000

dist_env:
  backend: nccl
  timeout_minutes: 30

model:
  _target_: nemo_automodel.NeMoAutoModelForCausalLM.from_config
  config:
    _target_: transformers.AutoConfig.from_pretrained
    pretrained_model_name_or_path: ./llama_1b_pythia
  torch_dtype: bf16
  trust_remote_code: True

checkpoint:
  enabled: True
  checkpoint_dir: path/to/checkpoints
  model_save_format: safetensors
  save_consolidated: False
  load_base_model: False

dataset:
  _target_: nemo_automodel.components.datasets.llm.megatron_dataset.MegatronPretraining
  paths: /path/to/dataset/pile_20B_tokenizer_text_document
  index_mapping_dir: /path/to/dataset/mapping_dir
  seq_length: 2048
  tokenizer:
    _target_: transformers.AutoTokenizer.from_pretrained
    pretrained_model_name_or_path: ./llama_1b_pythia
  split: "1.00, 0.00, 0.00" 
  splits_to_build: "train" 

dataloader:
  _target_: torchdata.stateful_dataloader.StatefulDataLoader
  shuffle: false
  collate_fn: torch.utils.data.default_collate

loss_fn:
  _target_: nemo_automodel.components.loss.masked_ce.MaskedCrossEntropy

optimizer:
  _target_: torch.optim.AdamW
  lr: 3.0e-4
  betas: [0.9, 0.95]
  weight_decay: 0.1

distributed:
  _target_: nemo_automodel.components.distributed.fsdp2.FSDP2Manager
  dp_size: none
  tp_size: 1
  cp_size: 1
  sequence_parallel: false

lr_scheduler:
  lr_decay_style: cosine
  min_lr: 3.0e-5 
  lr_warmup_steps: 1500
```

**Expected behavior**

The loss should converge smoothly, and grad_norm should remain stable (without periodic oscillations) throughout the training process.

**Additional context**

Environment & Logs

- nemo_automodel: 0.2.0 / 0.3.0 / 0.3.0rc0 (main)

- 0.2.0:

  - torch: 2.9.0+cu128 CUDA 12.8

  - transformers: 4.57.1

- 0.3.0rc0 (main):

  - torch: 2.10.0+cu128 CUDA 12.8

  - transformers: 5.3.0

**Comment by akoumpa** [comment]
Hi @ji-chengzhi ,

Thank you for reporting the issue.

One thing I would try is TE's optimizer
```
optimizer:
  _target_: transformer_engine.pytorch.optimizers.fused_adam.FusedAdam
  betas: [0.9, 0.95]
  eps: 1e-8
  lr: 1.0e-5
  weight_decay: 0
  master_weights: true
  store_param_remainders: true
  exp_avg_dtype: torch.float32
  exp_avg_sq_dtype: torch.float32
```
But I will also look into whether this is due to a deeper bug.

**Comment by ji-chengzhi** [comment]
Hi @akoumpa ,  just checking in to see if you’ve had a chance to reproduce the issue on your end? We are still quite intrigued by this since the behavior is indeed very strange.

Let us know if you were able to replicate the pattern or if you need any further information from our side. Thanks!

**Comment by akoumpa** [comment]
Hi @ji-chengzhi , I apologize for the delay, unfortunately, i had to prioritize bug fixing for our upcoming release, i hope to find some time this weekend. Overall, my hypotheses are (a) bf16 is losing precision small updates which are critical here and (b) the dataloader does not use shuffling which may be causing the repeated pattern (but this is speculation at this point). 

I need to find some time to run a few things on and report back -- Thank you for raising this, I have in mind to update the yaml recipe for pretraining for sure.

**Comment by qiaochuz-nv** [comment]
## QA RCCA Analysis

**Fix PR:** None — under investigation

**Root Cause:** Not confirmed as code defect. Community user reports periodic loss oscillations and sinusoidal grad_norm patterns when pre-training Llama 3.2 (1B/500M). Dev hypotheses: (a) bf16 precision losing small critical updates, (b) dataloader not shuffling. No conclusion reached.

**Trigger Config:** Pre-training + Llama 3.2 1B/500M + torch.optim.AdamW + bf16 + no shuffle + large GBS (2M tokens)

**Functional Test Coverage: N/A** — not a confirmed code defect.

**Conclusion:** Monitor for dev resolution. If confirmed as code bug with fix, re-assess for functional test coverage.

**Comment by akoumpa** [comment]
Hi @ji-chengzhi I'll try to reproduce today

**Comment by akoumpa** [comment]
Hi @ji-chengzhi , 

We are porting the default reduce dtype to float32 from https://github.com/MIND-AGI/Automodel/commit/95574ef7a1f23f2d5d5bf28f6d21025d9edba16f, in the main, and @yuhezhang-ai is working on additional validation on our side.

Thank you again :bow:

**Comment by ji-chengzhi** [comment]
Hi @akoumpa ,

Thank you for response and for your team's extra validation effort!

However, we would like to share some further insights from our deep dive. We believe that modifying the reduce_dtype alone (as done in MIND-AGI@95574ef) might not fundamentally resolve the root cause of this training instability.

Here is our core analysis:

**The Root Cause**: In the current default configuration, setting `model: torch_dtype: bf16` initializes the model weights entirely in bf16. When using the vanilla `torch.optim.AdamW`, the optimizer applies updates directly to these `bf16` parameters. This approach is overly aggressive for large-scale pre-training, as bf16 lacks the precision required to preserve small, critical gradient updates, leading to the periodic oscillations we observed.

**Why TE FusedAdam Works**: The reason TE's FusedAdam with `master_weights` stabilizes the training is that it maintains a dedicated copy of float32 master weights for the optimizer updates, preventing underflow during the optimization step.

**The Recommended Fix**: Instead of keeping the model parameters natively in bf16, the standard and more robust practice for FSDP pre-training should be:

- Keep the default model parameter initialization as `float32` (i.e., modifying or removing the default `torch_dtype: bf16` at the model level).

- Use the FSDP Mixed Precision Policy (`MixedPrecisionPolicy`) to cast the precision to `bf16` dynamically for computation (forward/backward passes) and communication, while maintaining `float32` master weights for the optimizer.

We hope this helps clarify the underlying behavior. Looking forward to your validation results!

Also, we noticed a recent PR https://github.com/NVIDIA-NeMo/Automodel/pull/2248 already changing this to `torch_dtype: float32`. Glad to see this direction, as we believe it's the fix for the stability issue we've been tracking..

**Comment by yuhezhang-ai** [comment]
Hi @ji-chengzhi, thanks for the detailed analysis. I totally agree with your diagnosis.

In the mentioned PR I added a mixed-precision guide and updated the pre-training recipe configs to use stable options: either `model.torch_dtype = fp32` with `torch.optim.AdamW`, or TE `FusedAdam` with `master_weights=True` and f32 for weights/states. One practical note is that memory usage varies by model between the two options, so users can pick whichever fits their workload. Thanks again for surfacing this.

I also ran additional validation including a full training run compared against an independent baseline. With the recommended settings, the training curve is smooth and converges to the same final loss.

**Comment by yuhezhang-ai** [comment]
Relevant precision bug fixes are already merged (such as https://github.com/NVIDIA-NeMo/Automodel/pull/2419), so functionality is correct and we closed this issue. 

I am opening a new issue to track the UX problem - default dtype if user didn't specify: https://github.com/NVIDIA-NeMo/Automodel/issues/2475
